src/specular/stages/batches.rs

```diff
@@ -218,20 +219,46 @@ fn decode_batches_v0(
     // Get l2 safe epoch info.
     let mut epoch_num = state.safe_epoch.number;
     let mut epoch_hash = state.safe_epoch.hash;
+    // Record the local l2 block number to check for duplicates and missing blocks.
+    let mut local_l2_num = safe_l2_num;
     // Decode each batch list in the batcher transaction.
     for batch_list in batch_lists.iter() {
         // Decode the first l2 block number at offset 0.
-        let first_l2_block_num: u64 = batch_list.val_at(0)?;
-        let first_l2_block_timestamp =
-            (first_l2_block_num - safe_l2_num) * config.chain.blocktime + safe_l2_ts;
+        let batch_first_l2_num: u64 = batch_list.val_at(0)?;
+        // Check for duplicates.
+        if batch_first_l2_num < local_l2_num {
+            tracing::warn!("invalid batcher transaction: contains already accepted batches | safe_head={} first_l2_block_num={}", local_l2_num, batch_first_l2_num);
+            eyre::bail!("invalid batcher transaction: contains already accepted batches");
```

Comment:
nit: would remove log here since duplicates the error (and add the context formatting to the error)

Reply:
the error gets ignored so that's why I add a warning